trpcultivate_germplasm/src/Plugin/TripalImporter/GermplasmAccessionImporter.php

```diff
@@ -0,0 +1,885 @@
+   *   The species of the organism.
+   * @param string $germplasm_subtaxa
+   *   Optional. Must consist of two strings, one of the subtaxon type
+   *   followed by the name. For example: "subspecies chadoii".
+   * @return int|false
+   *   The value of the primary key for the organism record in Chado.
+   *   If no single primary key can be retrieved, then FALSE is returned.
+   */
+  public function getOrganismID($genus_name, $germplasm_species, $germplasm_subtaxa) {
+
+    $organism_name = $genus_name . ' ' . $germplasm_species;
+    if ($germplasm_subtaxa) {
+      $organism_name = $organism_name . ' ' . $germplasm_subtaxa;
+    }
+    $organism_array = chado_get_organism_id_from_scientific_name($organism_name);
+
+    if (!$organism_array) {
+      $this->logger->error("Could not find an organism \"@organism_name\" in the database.", ['@organism_name' => $organism_name]);
+      $this->error_tracker = TRUE;
+      return false;
+    }
+    // We also want to check if we were given only one value back, as there is
+    // potential to retrieve multiple organism IDs
+    if (is_array($organism_array) && (count($organism_array) > 1)) {
+      $this->logger->error("Found more than one organism ID for \"@organism_name\" when only 1 was expected.", ['@organism_name' => $organism_name]);
+      $this->error_tracker = TRUE;
+      return false;
+    }
+
+    return $organism_array[0];
+  }
+
+  /**
+   * Checks if a stock exists in Chado and if not, inserts it and returns the primary
+   * key in the stock table. If the stock already exists, logs an error
+   *
+   * @param string $germplasm_name
+   *   The name of the germplasm.
+   * @param string $accession_number
+   *   A unique identifier for the germplasm accession.
+   * @param int $organism_id
+   *   The primary key of the stock's organism in the organism table
+   * @return int|false
+   *   The value of the primary key for the stock record in Chado. If the stock already
+   *   exists and does not match the accession number or type, or it cannot be inserted,
+   *   then FALSE is returned.
+   */
+  public function getStockID($germplasm_name, $accession_number, $organism_id) {
+
+    $accession_type_id = $this->getCVterm('accession');
+
+    // First query the stock table:
+    // 1. Using a regular condition to ensure the organism_id is a match
+    // 2. Create an OR condition group to look for records that match germplasm name OR 
+    //    the uniquename. Since the unique constraint is organism_id/uniquename/type_id,
+    //    we have to make sure this combo doesn't already exist with a different germplasm 
+    //    name.
+    $query = $this->connection->select('1:stock', 's')
+      ->fields('s', ['stock_id', 'name', 'uniquename', 'type_id'])
+      ->condition('s.organism_id', $organism_id, '=');
+
+    $orGroup = $query->orConditionGroup()
+      ->condition('s.name', $germplasm_name, '=')
+      ->condition('s.uniquename', $accession_number, '=');
```

Comment:
Looking at this is reminding me that we still need to have that discussion about uniquename in KP. Note: not requesting changes ;-)

**Recap / Brain dump**
Currently our uniquename/accession is KP:GERM[stock_id] which we cannot fill into the import template ahead of time since we do not know the stock_id until after insert. We have discussed that perhaps we should keep the primary accession as the original accession for that germplasm (i.e. genebank accession, cross number, RIL number) but what if we do not have an existing accession to apply (i.e. random seed from the grocery store)? Also, what about other sites which may want to stick with giving all germplasm their own unique accession? We'll want to walk through an example of how to get around this or test it to see what happens.

Reply:
Moved to issue #12